### DISCUSSION TITLE
Run error with comand of `message_ix.macro.calibrate()`

### DISCUSSION THREAD

**Discussion by my0828** [iiasa/message_ix#727] [MESSAGE modeling]
An error occurs when running `message_ix.macro.calibrate()`
Maybe the question is the equ of ` def gams_version():` in ixmp/model/gams.py.
The code and error message are as follows, what should I do.

> message_ix.macro.calibrate(scen, check_convergence=True)

> **CalledProcessErro**r: Command '['gams', 'null', '-LogOption=3']' returned non-zero exit status 1.

**Comment by glatterf42** [comment]
Hi @my0828, thanks for opening this discussion! Could I ask you to provide some more context, please? Please let us know what you want to do, so what is the code you want to run (even before `message_ix.macro.calibrate()` call)? And could you please also provide the output of running `message-ix show-versions` on the command line?

**Reply by my0828** [reply]
message-ix show-versions: 3.3.0
The previous code is as follows：

```python
model = 'Westeros Electrified'
base = message_ix.Scenario(mp, model=model, scenario="baseline")
scen = base.clone(model, 'Macro_calibrate','calibrate', keep_solution=False)

scen.check_out()

c=message_ix.macro.Calculate(s=base, data='westeros_macro_input.xlsx')
c.read_data()
c.derive_data()
message_ix.models.MACRO.initialize(scen)
message_ix.macro.add_model_data(base, scen, data='westeros_macro_input.xlsx')
message_ix.macro.calibrate(scen, check_convergence=True)
```

Erro message: File D:\Anaconda\lib\site-packages\ixmp\model\gams.py:34, in gams_version()
     31 gms.write_text("$exit;")
     33 # Execute, capturing stdout
---> 34 output = check_output(
     35     ["gams", "null", "-LogOption=3"],
     36     shell=os.name == "nt",
     37     cwd=tmp_dir,
     38     universal_newlines=True,
     39 )
     41 # Clean up
     42 gms.unlink()

> CalledProcessError: Command '['gams', 'null', '-LogOption=3']' returned non-zero exit status 1.

**Comment by glatterf42** [comment]
Thank you for the additional context. Could you please provide the full output of `message-ix show-versions`? It's not just about the version of message_ix, I am wondering whether your GAMS installation is being recognized by your system.
On that note, you can paste code in here using \`code\` and you can have it beautifully formatted on multiple lines using
\```python
some more code
\```
(or any other common language, even shell or bash for command line output).
Lastly, your version of message_ix is quite outdated (version [3.3.0](https://github.com/iiasa/message_ix/releases/tag/v3.3.0) was published more than two years ago), so please consider updating your code as this will make troubleshooting much easier and it might just resolve your error already!

**Reply by my0828** [reply]
the full output of message-ix show-versions as follows:
```C:\Users\mu;message-ix show-versions
'gams' 不是内部或外部命令，也不是可运行的程序
或批处理文件。

ixmp:        3.4.0 message_ix:  3.3.0 message_ix_models: None message_data: None

click:       8.0.4 dask:        2022.02.1 genno:       installed graphviz:    None jpype:       1.4.0
… JVM path:  C:\Program Files\Java\jre1.8.0_341\bin\server\jvm.dll openpyxl:    3.0.9 pandas:      1.4.2 pint:        0.19.2 xarray:      0.20.1 yaml:        6.0

iam_units:   installed jupyter:     installed matplotlib:  3.5.1 plotnine:    0.8.0 pyam:        1.3.1

GAMS:        'gams' executable not in PATH

python:      3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
python-bits: 64
OS:          Windows
OS-release:  10 machine:     AMD64 processor:   Intel64 Family 6 Model 142 Stepping 12, GenuineIntel byteorder:   little
LC_ALL:      None
LANG:        None
LOCALE:      ('Chinese (Simplified)_China', '936')```

**Reply by glatterf42** [reply]
Thank you for that! 
It is generally not recommended to have ixmp and message_ix installed with different versions. Even more so since both versions are already somewhat old. Since you seem to be using Anaconda, please try updating the package versions as described [here](https://stackoverflow.com/a/45197778). (If you are using pip instead, you might find [this question](https://stackoverflow.com/questions/2720014/how-to-upgrade-all-python-packages-with-pip) helpful.)

Finally, your output indeed shows that `'gams' executable [is] not in PATH`, which it needs to be in order for message_ix to run. Please follow the steps outlined in our [installation instructions](https://docs.messageix.org/en/latest/install.html#gams-required) to add GAMS to the `PATH` environment variable.

**Reply by my0828** [reply]
I have upgraded ixmp and message_ix pacaages with different versions., Simultaneously GAMS has been added to the path, both for system-wide and user-specific environments.

However,when I run `message_ix.macro.calibrate()`,the error is still reported.

```message_ix.macro.calibrate(scen, check_convergence=True)
at.ac.iiasa.ixmp.exceptions.IxException: at.ac.iiasa.ixmp.exceptions.IxException: This Scenario must be checked in before performing this operation!```

**Reply by my0828** [reply]
moer error message:
```File MsgScenario.java:1317, in at.ac.iiasa.ixmp.objects.MsgScenario.toGDX()

File TimeSeries.java:837, in at.ac.iiasa.ixmp.objects.TimeSeries.assertTimeseriesIsLockedInDB()

File TimeSeries.java:849, in at.ac.iiasa.ixmp.objects.TimeSeries.assertTimeSeriesIsLockedInDB()

Exception: Java Exception.```

**Reply by my0828** [reply]
message-ix show-versions
```ixmp:        999
message_ix:  999
message_ix_models: None
message_data: None

click:       8.0.4
dask:        2022.02.1
genno:       installed
graphviz:    None
jpype:       1.4.0
… JVM path:  C:\Program Files\Java\jre1.8.0_341\bin\server\jvm.dll
openpyxl:    3.0.9
pandas:      2.0.3
pint:        0.19.2
xarray:      0.20.1
yaml:        6.0

iam_units:   installed
jupyter:     installed
matplotlib:  3.5.1
plotnine:    0.8.0
pyam:        1.3.1

GAMS:        37.1.0

python:      3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
python-bits: 64
OS:          Windows
OS-release:  10
machine:     AMD64
processor:   Intel64 Family 6 Model 142 Stepping 12, GenuineIntel
byteorder:   little
LC_ALL:      None
LANG:        None
LOCALE:      ('Chinese (Simplified)_China', '936')```

**Reply by glatterf42** [reply]
Great, it looks like you added GAMS to your path successfully. And while I don't understand why `message-ix show-versions` does not show a version for ixmp and why the message-ix version is `999`, it also looks like your initial error message was resolved and shifted to a new one now. The old one was about GAMS not operating (correctly), the new one is an exception from within ixmp, so it would be really helpful for us to know the version of ixmp you are using. 

Going back to your initial comment about the code you are running, I formatted your answer so that the code snippet is more readable. Looking at it now, I find myself a bit confused over the order in which you are running the commands. Is this an official tutorial notebook that you are following? Or where did you get this code? 
To clarify: I haven't run `MACRO` myself yet, but I don't quite understand why you 

1. Create a scenario
2. Clone that scenario and check it out
3. Call `macro.Calculate` on the *first* scenario
4. Read and derive some data without storing it
5. Initialize `MACRO` (all-caps as opposed to the other `macro` spellings) for the first scenario
6. Add data for the second scenario (based on the first, I assume)
7. Calibrate the first scenario

Is this how the workflow is supposed to be?
